nexus/src/db/model.rs

```diff
@@ -352,6 +353,55 @@ pub struct Ipv6Net(pub external::Ipv6Net);
 NewtypeFrom! { () pub struct Ipv6Net(external::Ipv6Net); }
 NewtypeDeref! { () pub struct Ipv6Net(external::Ipv6Net); }
 
+impl Ipv6Net {
+    /// Generate a random subnetwork from this one, of the given prefix length.
+    ///
+    /// `None` is returned if:
+    ///
+    ///  - `prefix` is less than this address's prefix
+    ///  - `prefix` is greater than 128
+    pub fn random_subnet(&self, prefix: u8) -> Option<Self> {
+        use rand::RngCore;
+
+        if prefix < self.prefix() || prefix > 128 {
+            return None;
+        }
+        if prefix == self.prefix() {
+            return Some(*self);
+        }
+
+        // Generate a random address
+        let mut rng = rand::thread_rng();
+        let random =
+            u128::from(rng.next_u64()) << 64 | u128::from(rng.next_u64());
+
+        // Generate a mask for the new address.
+        //
+        // Left-shift will zero-fill. Even if we bit-OR with the original
+        // prefix, there may still be zeros in the "middle" if the original
+        // prefix is small and the new one large.
+        //
+        // So use a right-shift, to arrive at the desired prefix length.
+        //
+        // Safety: We've already checked above that the subtraction won't panic,
+        // since `prefix` <= 128.
+        let full_mask = u128::MAX >> (128 - prefix);
```

Comment:
Since this is a prefix, wouldn't you want it to be masking the *leading* ones?

For `prefix = 64`, this is: `0x0000000000000000ffffffffffffffff`.

But don't we want to be representing `0xffffffffffffffff0000000000000000` in this case?

This could be done with: `!(u128::MAX >> prefix)`; [example](https://doc.rust-lang.org/std/primitive.u128.html#method.leading_ones).